src/schema/object_identification.js

```diff
@@ -128,19 +126,25 @@ const NodeField = {
   },
   // Re-uses (slightly abuses) the existing GraphQL `resolve` function.
   resolve: (root, { __id }, request, rootValue) => {
-    const { type, id } = fromGlobalId(__id)
-    if (isSupportedType(type)) {
-      let exported = SupportedTypes.typeModules[type]
+    const { type: typeName, id } = fromGlobalId(__id)
+    if (isSupportedType(typeName)) {
+      let exported = SupportedTypes.typeModules[typeName]
       if (typeof exported === "function") {
         exported = exported()
       }
-      const { resolve } = exported
-      return resolve(
-        null,
-        argumentsForChild(type, id),
-        request,
-        rootValueForChild(rootValue)
-      )
+      const { resolve, type } = exported
+      return Promise.resolve(
+        resolve(
+          null,
+          argumentsForChild(typeName, id),
+          request,
+          rootValueForChild(rootValue)
+        )
+      ).then(data => {
+        // Add the already known type so `NodeInterface` can pluck that out in
+        // its `resolveType` implementation.
+        return { __type: type, ...data }
```

Comment:
Why not continue to use `__typename` here, like you did in the `ConversationItem` union, (and then update to `resolveType({ __typename })` )?

Reply:
Here I’m actually passing the materialised type object, not its name. Otherwise `resolveType` has to do the above dance to fetch `type` again, which felt a bit all over the place.